vnext/ReactUWP/Views/ViewControl.cpp

```diff
@@ -5,94 +5,252 @@
 
-ViewControl::ViewControl()
+DynamicAutomationPeer::DynamicAutomationPeer(winrt::FrameworkElement const& owner)
+  : Super(owner)
 {
 }
 
-/*static*/ winrt::com_ptr<ViewControl> ViewControl::Create()
+winrt::hstring DynamicAutomationPeer::GetClassNameCore() const
 {
-  return winrt::make_self<ViewControl>();
+  return winrt::to_hstring("DynamicAutomationPeer");
 }
 
-winrt::Windows::UI::Xaml::Automation::Peers::AutomationPeer ViewControl::OnCreateAutomationPeer()
+winrt::AutomationControlType DynamicAutomationPeer::GetAutomationControlTypeCore() const
 {
-  auto dyn = winrt::make<winrt::react::uwp::implementation::DynamicAutomationPeer>(*this);
-  return dyn;
+  if (auto viewControl = GetViewControl())
+  {
+    switch (viewControl->AccessibilityRole())
+    {
+    case ::react::uwp::AccessibilityRoles::Button:
+    case ::react::uwp::AccessibilityRoles::ImageButton:
+      return winrt::AutomationControlType::Button;
+    case ::react::uwp::AccessibilityRoles::Link:
+      return winrt::AutomationControlType::Hyperlink;
+    case ::react::uwp::AccessibilityRoles::Image:
+      return winrt::AutomationControlType::Image;
+    case ::react::uwp::AccessibilityRoles::KeyboardKey:
+      return winrt::AutomationControlType::Custom;
+    case ::react::uwp::AccessibilityRoles::Text:
+    case ::react::uwp::AccessibilityRoles::Summary:
+    case ::react::uwp::AccessibilityRoles::Header:
+      return winrt::AutomationControlType::Text;
+    case ::react::uwp::AccessibilityRoles::Adjustable:
+      return winrt::AutomationControlType::Slider;
+    case ::react::uwp::AccessibilityRoles::Search:
+    case ::react::uwp::AccessibilityRoles::Unknown:
+    default:
+      return winrt::AutomationControlType::Group;
+    }
+  }
+
+  return winrt::AutomationControlType::Group;
 }
 
-} } // namespace react::uwp
+winrt::IInspectable DynamicAutomationPeer::GetPatternCore(winrt::PatternInterface const& patternInterface) const
+{
+  if (auto viewControl = GetViewControl())
+  {
+    auto accessibilityRole = viewControl->AccessibilityRole();
 
-namespace winrt::react::uwp::implementation
+    if (patternInterface == winrt::PatternInterface::Invoke &&
+      (accessibilityRole == ::react::uwp::AccessibilityRoles::Button || accessibilityRole == ::react::uwp::AccessibilityRoles::ImageButton))
+    {
+      return *this;
+    }
+    else if (patternInterface == winrt::PatternInterface::Selection || patternInterface == winrt::PatternInterface::SelectionItem)
+    {
+      return *this;
+    }
+  }
+
+  return Super::GetPatternCore(patternInterface);
+}
+
+bool DynamicAutomationPeer::IsEnabledCore() const
 {
-    DynamicAutomationPeer::DynamicAutomationPeer(Windows::UI::Xaml::FrameworkElement const& owner)
-      : Super(owner)
+  bool disabled = false;
+
+  try
+  {
+    if (auto viewControl = GetViewControl())
     {
+      disabled = viewControl->AccessibilityState(::react::uwp::AccessibilityStates::Disabled);
     }
+  }
+  catch (winrt::hresult_error const & ex) {}
```

Comment:
Why exception here? If it's because viewControl memory is freed, this code would be very dangerous, and I prefer not to ship this feature until the problem is resolved.
Even if it's not crash in debug version, how about release version?